### ISSUE TITLE
Flags created with --define appear to be incorrectly cached

### ISSUE THREAD

**Issue by hilcode** [com-lihaoyi/mill#3789]
Given this `build.sc`:

```scala
import mill.T

def myProperty: T[Boolean] = T.input {
    println(s"""System.getProperty("my-property", "X")=${System.getProperty("my-property", "X")}""")
    System.getProperty("my-property", "X") match {
        case result if result == "F" || result == "X" =>
            println(s"$result ::: my-property=false")
            false

        case "T" =>
            println("T ::: my-property=true")
            true

        case x =>
            println(s"my-property='$x'")
            throw new IllegalStateException("Invalid value")
    }
}

def foo: T[String] = T {
    if (myProperty()) {
        println("TRUE")
        "TRUE"
    } else {
        println("FALSE")
        "FALSE"
    }
}
```

I see the following:

```scala
~/c/s/mill-bug> ./mill --version
Mill Build Tool version 0.11.12
Java version: 21.0.1, vendor: Eclipse Adoptium, runtime: /nix/store/6m2wff2lm5pk892389ivgcbs1nqbsz6v-temurin-bin-21.0.1
Default locale: en_US, platform encoding: UTF-8
OS name: "Linux", version: 6.11.3-1-default, arch: amd64

~/c/s/mill-bug> rm -rf out

~/c/s/mill-bug> ./mill --define=my-property=T foo
[build.sc] [54/59] compile
[info] compiling 1 Scala source to .../out/mill-build/compile.dest/classes ...
[info] done compiling
[1/2] myProperty
System.getProperty("my-property", "X")=T
T ::: my-property=true
[2/2] foo
TRUE

~/c/s/mill-bug> ./mill foo
[1/2] myProperty
System.getProperty("my-property", "X")=T
T ::: my-property=true

~/c/s/mill-bug> ./mill --define=my-property=F foo
[1/2] myProperty
System.getProperty("my-property", "X")=F
F ::: my-property=false
[2/2] foo
FALSE

~/c/s/mill-bug> ./mill foo
[1/2] myProperty
System.getProperty("my-property", "X")=F
F ::: my-property=false
```
Note that when _not_ providing `--define=my-property=?`, it seems that the previous invocation of `myProperty` is reused somehow, including the `println` output.

**Comment by lefou** [comment]
@lihaoyi just fixed this issue in `main` branch, which will become `0.12.0`. We might want to backport the fix to `0.11.x` too.

**Comment by lefou** [comment]
Backported to `0.11.x` branch https://github.com/com-lihaoyi/mill/pull/3795 and eventually released in `0.11.13`.
